### PR TITLE
Add gated histogram quantile balancing

### PR DESCRIPTION
**Pull request by rkarimimahab**
Enable the Kimi K3-style whole-step histogram implementation with:

--moe-router-load-balancing-type quantile_balancing

--moe-router-quantile-balancing-histogram

--moe-router-score-function sigmoid

--moe-aux-loss-coeff 0

- [ ] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/training/arguments.py [hunk 1]
```diff
@@ -3534,7 +3534,13 @@ def _add_moe_args(parser):
...
+                       help='Determines the load balancing strategy for the router. "aux_loss" corresponds to the load balancing loss used in GShard and SwitchTransformer; "seq_aux_loss" corresponds to the loss used in DeepSeekV2; "sinkhorn" corresponds to S-BASE; "quantile_balancing" (QB) updates a routing bias without an auxiliary loss, with the Kimi K3 histogram variant separately gated; "none" implies no load balancing. The default is "aux_loss".')
+    group.add_argument('--moe-router-quantile-balancing-ema', type=float, default=0.0,
+                       help='EMA coefficient for the whole-step quantile-balancing expert bias.')
+    group.add_argument('--moe-router-quantile-balancing-num-bins', type=int, default=1000,
+                       help='Number of per-expert histogram bins used by whole-step quantile balancing.')
+    group.add_argument('--moe-router-quantile-balancing-histogram', action='store_true',
+                       help='Opt in to Kimi K3-style whole-step histogram quantile balancing.')
```
Comment (Mellonta) [inline megatron/training/arguments.py:3543]: Duplicate CLI registration?

File: megatron/core/transformer/moe/router.py [hunk 2]
```diff
@@ -335,48 +365,26 @@ def quantile_balancing(self, logits: torch.Tensor) -> tuple[torch.Tensor, torch.
             self.config.moe_router_num_groups is None and self.config.moe_router_group_topk is None
         ), "Quantile balancing routing does not support group-limited routing."
 
-        local_num_tokens = logits.shape[0]
-        # Gather logits across TP/CP so the quantile sees a whole sequence's tokens.
-        # The DP reduction and qb_beta update run at the global-batch boundary in
-        # finalize_model_grads._update_router_qb_beta.
-        gather_group = self.tp_cp_group
```
Comment (Mellonta) [inline megatron/core/transformer/moe/router.py:342]: Initially I implemented non-histogram QB by computing an exact quantile per microbatch and averaging across microbatches and DP replicas. Now when histogram=off it instead computes rank-local quantiles and averages across TP/CP as well, so the histogram=off path may be unusable. I am thinking of either preserving the original non-histogram implementation or simply removing it and always use histogram QB (if maintaining two is tedious).

File: megatron/core/transformer/moe/router.py [hunk 3]
```diff
@@ -335,48 +365,26 @@ def quantile_balancing(self, logits: torch.Tensor) -> tuple[torch.Tensor, torch.
...
+                scores = torch.sigmoid(logits.detach().float())
+                if should_update_bias:
+                    histogram_scores = scores if padding_mask is None else scores[~padding_mask]
+                    self.local_quantile_balancing_histogram += compute_quantile_balancing_histogram(
+                        histogram_scores,
+                        self.qb_beta,
+                        self.topk,
+                        num_bins=self.config.moe_router_quantile_balancing_num_bins,
+                    )
+                indices = (scores + self.qb_beta).topk(self.topk, dim=1).indices
```
Comment (Mellonta) [inline megatron/core/transformer/moe/router.py:380]: Did we want to change from (scores - self.qb_beta) to (scores + self.qb_beta) here? We might also need to change the inference side in TopKRouter._forward, which is still using -.

File: megatron/core/distributed/finalize_model_grads.py [hunk 4]
```diff
@@ -374,41 +379,71 @@ def _update_router_expert_bias(
...
+        ) * stacked_local_average
+        updated -= updated.mean(dim=-1, keepdim=True)
+        for qb_beta, new_beta in zip(qb_beta_list, updated):
+            qb_beta.copy_(new_beta)
+        return
 
-    Averages each router's accumulated quantile (qb_beta_accum/qb_beta_count) across
-    DP, EMA-blends it with the current qb_beta, re-centers, and writes it back.
-    """
+    """Pool whole-step QB histograms and update the per-expert bias once per global batch."""
```
Comment (Mellonta) [inline megatron/core/distributed/finalize_model_grads.py:416]: Wrong place?

File: megatron/core/transformer/moe/moe_utils.py [hunk 5]
```diff
@@ -245,6 +245,93 @@ def qb_dual_update(
...
+        if topk >= num_experts:
+            raise ValueError("Quantile Balancing requires topk < num_experts.")
+        if num_bins <= 0 or score_span <= 0.0:
+            raise ValueError("Quantile Balancing requires positive histogram bins and score span.")
+
+        histogram = torch.zeros((num_experts, num_bins), dtype=torch.int32, device=scores.device)
+        if num_tokens == 0:
+            return histogram
+
+        cutoff = (scores + expert_bias).kthvalue(num_experts - topk, dim=1).values.unsqueeze(1)
```
Comment (Mellonta) [inline megatron/core/transformer/moe/moe_utils.py:269]: Two selection passes over the same tensor: `kthvalue` for the cutoff here, `topk` for the indices in router.py. Using `topk + 1` and we can get both in one pass, last value is the cutoff, rest are the indices. Same as the original `qb_dual_update`.
